### PR TITLE
DEV: Modernizing code to Glimmer components and the new modal API

### PR DESCRIPTION
**Pull request by megothss**

### CODE REVIEW

File: assets/javascripts/discourse/initializers/d-templates.js [hunk 1]
```diff
@@ -79,12 +80,9 @@ function addKeyboardShortcut(api, container) {
             textarea: activeElement,
           });
         } else {
-          const modalController = container.lookup(
-            "controller:discourse-templates-modal"
-          );
-
-          showModal("discourse-templates-modal");
-          modalController.set("textarea", activeElement);
+          getOwner(this)
```
Comment (CvX) [inline assets/javascripts/discourse/initializers/d-templates.js:83]: I don't know what `this` is in this context. It's probably better to use the available `container` to do the `lookup()`

File: assets/javascripts/discourse/components/d-templates/modal/form.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    this.#textarea = this.args.model?.textarea;
+  }
+
+  get dataModel() {
+    return this.#dataModel;
+  }
+
+  get textarea() {
+    return this.#textarea;
+  }
```
Comment (CvX) [inline assets/javascripts/discourse/components/d-templates/modal/form.js:23]: To keep things simple - I think this works too

```suggestion
  dataModel =
    this.args.model?.model ||
    getOwner(this).lookup("controller:composer").model;
  #textarea = this.args.model?.textarea;
```

File: assets/javascripts/discourse/components/d-templates/item.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { action } from "@ember/object";
+import Component from "@glimmer/component";
+
```
Comment (CvX) [inline assets/javascripts/discourse/components/d-templates/item.js:3]: 🧹 

```suggestion
```

File: assets/javascripts/discourse/components/d-templates/filterable-list.hbs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<div class="templates-filterable-list" {{did-insert this.load}}>
+
+  <ConditionalLoadingSpinner @condition={{this.loading}}>
+    <div class="templates-filter-bar">
+      {{#if this.siteSettings.tagging_enabled}}
+        <DTemplates::TagDrop
+          @availableTags={{this.availableTags}}
+          @tagId={{this.selectedTag}}
+          @onChangeSelectedTag={{action "changeSelectedTag"}}
```
Comment (CvX) [inline assets/javascripts/discourse/components/d-templates/filterable-list.hbs:9]:
```suggestion
          @onChangeSelectedTag={{this.changeSelectedTag}}
```

File: assets/javascripts/discourse/components/d-templates/filterable-list.hbs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        @value={{this.listFilter}}
+        placeholder={{i18n "templates.filter_hint"}}
+      />
+    </div>
+    <div class="templates-list">
+      {{#each this.filteredReplies as |r|}}
+        <DTemplates::Item
+          @template={{r}}
+          @model={{@model}}
+          @onInsertTemplate={{action "insertTemplate"}}
```
Comment (CvX) [inline assets/javascripts/discourse/components/d-templates/filterable-list.hbs:23]:
```suggestion
          @onInsertTemplate={{this.insertTemplate}}
```

File: assets/javascripts/discourse/components/d-templates/tag-drop.js [hunk 6]
```diff
@@ -35,13 +35,12 @@ export default TagDrop.extend({
...
-      // the property onChange in the handlebars template
-      this.onChange && this.onChange(tagId, tag);
-    },
-  },
-});
+  @action
+  onChange(tagId, tag) {
+    // overrides the action onChange of the parent with the value received in
+    // the property onChangeSelectedTag in the handlebars template
+    this.onChangeSelectedTag && this.onChangeSelectedTag(tagId, tag);
```
Comment (CvX) [inline assets/javascripts/discourse/components/d-templates/tag-drop.js:44]: 🏌️ 
```suggestion
    this.onChangeSelectedTag?.(tagId, tag);
```

File: assets/javascripts/lib/apply-template.js [hunk 7]
```diff
@@ -47,16 +49,16 @@ export function prepareTemplate(templateTitle, templateContent, model) {
...
 ) {
-  const model = context.get("model");
-
   // insert the title if blank
   if (model && !model.title) {
     model.set("title", templateTitle);
   }
 
   // insert the content of the template in the compose
-  context.appEvents.trigger("composer:insert-block", templateContent);
```
Comment (CvX) [inline assets/javascripts/lib/apply-template.js:61]: The previous pattern of passing `appEvents` in was better. `getOwner` in this case relies on our hack that falls back to basically a global variable when used with an object that doesn't have an owner (`this` in a standalone function)

File: assets/javascripts/discourse/initializers/d-templates.js (hunk 1 above)
Comment (megothss) [reply assets/javascripts/discourse/initializers/d-templates.js]: Done

File: assets/javascripts/lib/apply-template.js (hunk 7 above)
Comment (megothss) [reply assets/javascripts/lib/apply-template.js:61]: I'd like to skip this change.

I have a follow up PR that extract some logic into a service and will change this line:

https://github.com/discourse/discourse-templates/compare/templates-2023...dev/templates-service

**Review by CvX — APPROVED** [review]

**Pull request merged**